gpflow/training/natgrad_optimizer.py

```diff
@@ -0,0 +1,322 @@
+
+        # these are the tensorflow variables to assign
+        q_mu_u = q_mu_param.unconstrained_tensor
+        q_sqrt_u = q_sqrt_param.unconstrained_tensor
+
+        # so the transform to work for LowerTriangular
+        mean_new.set_shape(q_mu_param.shape)
+        varsqrt_new.set_shape(q_sqrt_param.shape)
+
+        q_mu_assign = tf.assign(q_mu_u, q_mu_param.transform.backward_tensor(mean_new))
+        q_sqrt_assign = tf.assign(q_sqrt_u, q_sqrt_param.transform.backward_tensor(varsqrt_new))
+        return q_mu_assign, q_sqrt_assign
+
+#
+# Xi transformations necessary for natural gradient optimizer.
+# Abstract class and two implementations: XiNat and XiSqrtMeanVar.
+#
+
+class XiTransform(metaclass=abc.ABCMeta):
+    """
+    XiTransform is the base class that implements three transformations necessary
+    for the natural gradient calculation wrt any parameterization.
+    This class does not handle any shape information, but it is assumed that
+    the parameters pairs are always of shape (N, D) and (N, N, D).
+    """
+    @abc.abstractmethod
+    def meanvarsqrt_to_xi(self, mean, varsqrt):
+        """
+        Transforms the parameter `mean` and `varsqrt` to `xi_1`, `xi_2`
+
+        :param mean: the mean parameter (N, D)
+        :param varsqrt: the varsqrt parameter (N, N, D)
+        :return: tuple (xi_1, xi_2), the xi parameters (N, D), (N, N, D)
+        """
+        pass
+
+    @abc.abstractmethod
+    def xi_to_meanvarsqrt(self, xi_1, xi_2):
+        """
+        Transforms the parameter `xi_1`, `xi_2` to `mean`, `varsqrt`
+
+        :param xi_1: the xi_1 parameter
+        :param xi_2: the xi_2 parameter
+        :return: tuple (mean, varsqrt), the meanvarsqrt parameters
+        """
+        pass
+
+    @abc.abstractmethod
+    def naturals_to_xi(self, nat_1, nat_2):
+        """
+        Applies the transform so that `nat_1`, `nat_2` is mapped to `xi_1`, `xi_2`
+
+        :param nat_1: the nat_1 parameter
+        :param nat_2: the nat_1 parameter
+        :return: tuple `xi_1`, `xi_2`
+        """
+        pass
+
+
+class XiNat(XiTransform):
+    def meanvarsqrt_to_xi(self, mean, varsqrt):
+        return meanvarsqrt_to_natural(mean, varsqrt)
+
+    def xi_to_meanvarsqrt(self, xi_1, xi_2):
+        return natural_to_meanvarsqrt(xi_1, xi_2)
+
+    def naturals_to_xi(self, nat_1, nat_2):
+        return nat_1, nat_2
+
+
+class XiSqrtMeanVar(XiTransform):
+    def meanvarsqrt_to_xi(self, mean, varsqrt):
+        return mean, varsqrt
+
+    def xi_to_meanvarsqrt(self, xi_1, xi_2):
+        return xi_1, xi_2
+
+    def naturals_to_xi(self, nat_1, nat_2):
+        return natural_to_meanvarsqrt(nat_1, nat_2)
+
+
+#
+# Auxiliary gaussian parameter conversion functions.
+# 
+# The following functions expect their first and second inputs to have shape
+# DN1 and DNN, respectively. Return values are also of shapes DN1 and DNN.
+
+
+def swap_dimensions(method):
+    """
+    Converts between GPflow indexing and tensorflow indexing
```

Comment:
Is this still necessary after @markvdw's rolling of dimensions in #612  ?

Reply:
Yes, for the mean only now though (if the mean is KM1 and the var is KMM then all the matrix ops work out nicely). It could be written without, but at the expense of being slightly harder to read

Reply:
I don't see the purpose of this code anymore, since it is only called for the mean. Is it expected that two transposes and a matmul are more efficient than an einsum? If not then I would just remove that and use einsums